test_plans/is_device_copyable.asciidoc

```diff
@@ -0,0 +1,81 @@
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Types coverage
+
+All of the tests described below are performed using `typename T` which satisfies the https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable[requriments] of a device copyable type:
+
+* The application defines the trait `is_device_copyable_v<T>` to `true`;
+* Type `T` has at least one eligible copy constructor, move constructor, copy assignment operator, or move assignment operator;
+* Each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is public;
+* When doing an inter-device transfer of an object of type `T`, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;
+* Type `T` has a public non-deleted destructor;
+* The destructor has no effect when executed on the device.
+
+Define a new class of type `T` with public default constructor, destructor, copy constructor, copy assignment operator and several public members:
+```
+class T {
+  public:
+    T() = default;
+    ~T() = default;
+    T(T& t) : a(t.a), b(t.b), c(t.c) {}
+
+    T& operator=(const T& t) {
+        a = t.a;
+        b = t.b;
+        c = t.c;
+    }
+
+    int a;
+    float b;
+    char c;
+};
+```
+
+== Tests
+
+=== Cpp17UnaryTrait requirements.
+
+`is_device_copyable` must meet the https://en.cppreference.com/w/cpp/named_req/UnaryTypeTrait[Cpp17UnaryTrait] requirements:
+
+    * DefaultConstructible and CopyConstructible. Verify that:
+
+        ** `std::is_default_constructible_v<is_device_copyable<T>> == true`
+        ** `std::is_copy_constructible_v<is_device_copyable<T>> == true`
+
+    * Takes one template type parameter (additional template parameters are optional and allowed). Verify that: `std::is_constructible_v<is_device_copyable<T>> == true`
+
+    * Publicly and unambiguously derived from a specialization of integral_constant, known as its base characteristic. Verify that +
+    `std::is_base_of_v<std::integral_constant<bool, true>, is_device_copyable<T>> == true` or
+    `std::is_base_of_v<std::integral_constant<bool, false>, is_device_copyable<T>> == true`.
+
+    * The member names of the base characteristic (`std::integral_constant`) are not hidden and are unambiguously available. Verify that:
+
+        ** `std::integral_constant<bool, true>::value_type` and `is_device_copyable<T>::value_type` are the same type
+
+        ** `std::integral_constant<bool, true>::type` and `is_device_copyable<T>::type` are the same type
+        ** Operators `std::integral_constant<bool, true>::value_type() and `is_device_copyable<T>::value_type()` have the same return type
+        ** `std::integral_constant<bool, true>::operator()` returns true
```

Comment:
Isn't this last bullet item just testing something from the C++ headers?  It doesn't seem like it's testing anything from SYCL.

Reply:
It tests UnaryTrait requirement for `is_device_copyable`,  "The member names of the base characteristic are not hidden and are unambiguously available", therefore `is_device_copyable` must has access to `std::integral_contant` members. This is what this part is about, but I don't know how to test an access to members in other way. If it's not needed I can remove this part.

Reply:
If you want to test that `std::integral_constant::operator()` is unambiguously available to `is_device_copyable`, then it seems like you should test this instead:

* `is_device_copyable<T>::operator()` return true.

Reply:
Instead of all members of `std::integral_constant` or `operator()` only?

Reply:
It seems like your last bullet point:

> ** `std::integral_constant<bool, true>::operator()` returns true

is trying to validate whether `std::integral_constant::operator()` is unambiguously available to `is_device_copyable`.  If this is your intent, it seems like that bullet should be this instead:

> ** `is_device_copyable<T>::operator()` return true.

Reply:
Updated, but with comparison between them because it haven't always return true